lib/VolSurface/Calibration/Equities.pm

```diff
@@ -0,0 +1,637 @@
+    my $lambda2 = ((($num_of_points + 1)**0.5) - 1) / ($num_of_points * (2**0.5)) * $self->_scale;
+
+    my $i;
+    my $j;
+
+    #initialize the simplex with the initial guess
+    my @simplex = ($params);
+
+    my $step = 0;
+    for ($i = 0; $i < $num_of_var; $i++) {
+        $step = 0.1 * abs($params->[$i])
+            if (0.1 * abs($params->[$i]) > $step);
+    }
+
+    for (my $i = 1; $i <= $num_of_points; $i++) {
+        my @new = map { ($i - 1 == $_) ? $simplex[0][$_] + $step : 0 + $simplex[0][$_] } (0 .. $num_of_var - 1);
+        push @simplex, \@new;
+    }
+
+    my @function_eval = map { $self->function_to_optimize($_) } @simplex;
+    my $starting_value = $function_eval[0];
+
+    my @current_reflection;
+    my @current_expansion;
+    my @current_contraction;
+
+    my $expansion_function_eval;
+    my $reflected_function_eval;
+    my $contraction_function_eval;
+    my $lowest_point_index = 0;
+
+    my $calcvert = 1;
+    my $counter  = 0;
+
+    while (1) {
+        if ($calcvert == 1) {
+            $calcvert = 0;
+            for ($j = 0; $j <= $num_of_points; $j++) {
+                if ($j != $lowest_point_index) {
+                    for ($i = 0; $i < $num_of_points; $i++) {
+                        $current_reflection[$i] = $simplex[$j][$i];
+                    }
+                    $function_eval[$j] = $self->function_to_optimize(\@current_reflection);
+                }
+            }
+
+        }
+        my @sorted_eval = sort { $a <=> $b } @function_eval;
+        $lowest_point_index = first_index { $_ == $sorted_eval[0] } @function_eval;
+        my $highest_point_index = first_index { $_ == $sorted_eval[-1] } @function_eval;
+        my @indexes = indexes { $_ == $sorted_eval[-2] } @function_eval;
+
+        my @simplex_centroid = _calculate_simplex_centroid(\@simplex, $num_of_points, $num_of_var, $highest_point_index);
+
+        $highest_value = $function_eval[$highest_point_index];
+
+        my $convtol = $intol * ($starting_value + $intol);
+
+        last
+            if ($function_eval[$lowest_point_index] < $intol
+            or ($function_eval[$highest_point_index] < ($function_eval[$lowest_point_index] + $convtol))
+            or $counter > $self->_max_iteration);
+
+        #imple reflection of the highest point
+        @current_reflection =
+            map { $simplex_centroid[$_] + 1 * ($simplex_centroid[$_] - $simplex[$highest_point_index][$_]) } (0 .. $num_of_var - 1);
+        my $reflected_function_eval = $self->function_to_optimize(\@current_reflection);
+        if ($reflected_function_eval < $function_eval[$lowest_point_index]) {
+
+#Do simple expansion or in other words look up a little further in this direction
+            @current_expansion = map { $simplex_centroid[$_] + 2 * ($current_reflection[$_] - $simplex_centroid[$_]) } (0 .. $num_of_var - 1);
+
+            @simplex_centroid =
+                map { $current_reflection[$_] } (0 .. $num_of_var - 1);
```

Comment:
This is again a simple copy operation:

```perl
@simplex_centroid = @current_reflection;
```

Not sure, but this is called in a loop. Maybe it's better to make this simplex_centroid an array reference instead of an array. That way such swap operations can be much faster `$simplex_centroid = \@current_reflection`. Though, you have to be aware then that further changes in `@current_reflection` also change the centroid.